felix/deltatracker/delta_tracker.go

```diff
@@ -0,0 +1,335 @@
+	// As we iterate over the new values, we add them to the newXXX maps and delete them from the old.
+	// Then, at the end, the oldXXX maps will contain only the values that are now missing from the
+	// dataplane, we then handle those at the bottom.
+	oldInDPDesired := c.inDataplaneAndDesired
+	newInDPDesired := make(map[K]V)
+	oldInDPNotDesired := c.inDataplaneNotDesired
+	newInDPNotDesired := make(map[K]V)
+
+	err := iter(func(k K, v V) {
+		// Figure out if we _want_ it to exist and tee up update/deletion accordingly.
+		if desiredV, desired := c.GetDesired(k); desired {
+			// Record that this key exists in the new copy of the cache.
+			newInDPDesired[k] = v
+
+			// Check if the value is correct,
+			if c.valuesEqual(desiredV, v) {
+				// Value in dataplane is correct, clean up any pending update.
+				delete(c.desiredUpdates, k)
+			} else {
+				// Value in dataplane is incorrect.  Queue up an update (if there isn't one already).
+				c.desiredUpdates[k] = desiredV
+			}
+		} else {
+			// We don't want this key, but it's in the dataplane.  Queue up deletion.
+			newInDPNotDesired[k] = v
+		}
+
+		// Remove the key from the old cache, we'll then scan the old cache to find keys that are missing.
+		delete(oldInDPDesired, k)
+		delete(oldInDPNotDesired, k)
+	})
+	if err != nil {
+		// We may have failed mid-iteration, if we just returned an error here, we'd leave our internal
+		// state broken because we've removed all the keys that we've seen from oldInDPDesired and
+		// oldInDPNotDesired, and updated c.desiredUpdates to match the new KV's from the iterator.
+		// Fix that up by applying the new keys to the old in-DP maps:
+		for k, v := range newInDPDesired {
+			oldInDPDesired[k] = v
+		}
+		for k, v := range newInDPNotDesired {
+			oldInDPNotDesired[k] = v
+		}
+		return fmt.Errorf("failed to iterate over dataplane state: %w", err)
+	}
+
+	// oldInDPDesired now only contains KVs that we _thought_ were in the dataplane but are now gone.
+	for k := range oldInDPDesired {
+		if desiredV, desired := c.GetDesired(k); desired {
+			// We want this key, but it's missing, queue up an add.
+			c.desiredUpdates[k] = desiredV
+		} // else we don't want it, and it's gone; nothing to do.
+		delete(oldInDPDesired, k)
+	}
+
+	// Now done with oldInDPDesired, replace it.
+	c.inDataplaneAndDesired = newInDPDesired
+	c.inDataplaneNotDesired = newInDPNotDesired
+	logrus.WithFields(logrus.Fields{
+		"totalNumInDP":          len(c.inDataplaneAndDesired),
+		"desiredUpdates":        len(c.desiredUpdates),
+		"inDataplaneNotDesired": len(c.inDataplaneNotDesired),
+	}).Debug("Updated dataplane state.")
+
+	return nil
+}
+
+// SetDataplane updates a key in the dataplane cache.  I.e. it tells this tracker that the dataplane
+// has the given KV.  Updated the pending update/deletion set if the new KV differs from what is
+// desired.
+func (c *DeltaTracker[K, V]) SetDataplane(k K, v V) {
```

Comment:
Correct, all this method does is tell this object that "this key is in the dataplane".  Whether the key is _wanted_ has to be set via the `SetDesired()` method. So, you could use this to tell this object that you updated the dataplane, or that you detected an update from some other source.  For example, if we used this object to keep track of interfaces, we get netlink events for "added/removed/changed"; we could feed those into SetDataplane/DeleteDataplane.

Reply:
WDYT to the naming?  I wondered about doing an API like this:
```
dt.Desired().Set(x, y) / Get(x) 
dt.Dataplane().Set(x, y) / Get(x)
```
so that the API really emphasises that this is two maps that you manipulate

Reply:
I gave the above a go in the latest commit.  I think it comes out a lot cleaner from an API point of view.  Now the two maps can be seen and manipulated with a "typical" custom map API and, by replicating the approach on CachingMap, exposing the methods becomes a pass-through.  I used an interface to avoid exporting the delta tracker methods that don't make sense.